[JSC] Replace `isValidISODate` with `isDateTimeWithinLimits` in `TemporalCalendar::monthDayFromFields`

sosukesuzuki opened the pull request:
#### a1e2020dffa247ed280569b4f35dfaf230e75948
<pre>
[JSC] Replace `isValidISODate` with `isDateTimeWithinLimits` in `TemporalCalendar::monthDayFromFields`
<a href="https://bugs.webkit.org/show_bug.cgi?id=303163">https://bugs.webkit.org/show_bug.cgi?id=303163</a>

Reviewed by Yusuke Suzuki.

The `isValidISODate` check after `regulateISODate` is redundant because:

- When overflow is `Reject`, `regulateISODate` already calls `isValidISODate`
  and returns `std::nullopt` if invalid.
- When overflow is `Constrain`, the month and day are clamped to valid ranges,
  so `isValidISODate` will always return true.

However, according to the spec, `monthDayFromFields` should call
`ISODateWithinLimits` to check that the year is within range.
This is necessary for future non-ISO calendar support where the
reference year may be invalid in rare cases.

See: <a href="https://github.com/tc39/proposal-temporal/issues/2997">https://github.com/tc39/proposal-temporal/issues/2997</a>

* Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
(JSC::TemporalCalendar::monthDayFromFields):

Canonical link: <a href="https://commits.webkit.org/303878@main">https://commits.webkit.org/303878@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/85c5702dd9b5e777c16f78c8755aa7d8728d2757

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133863 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6374 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45054 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141440 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/85923 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/155/builds/6903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/156/builds/6238 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102400 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/85923 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136810 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/155/builds/6903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/120006 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83199 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/155/builds/6903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-API-Tests-EWS "Waiting to run tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/125940 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/155/builds/6903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38123 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144086 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/132377 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6044 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38702 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/110769 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/6126 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/156/builds/6238 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110970 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28143 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116261 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/59791 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6096 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/34536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/165340 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/5942 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69560 "Built successfully") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/43189 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/6187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6050 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

catamorphism reviewed (approved):
You're right about this check. According to the spec, https://tc39.es/proposal-temporal/#sec-temporal-calendarmonthdayfromfields , monthDayFromFields should actually be calling ISODateWithinLimits (which checks that the year is within range), not isValidISODate (which only checks the month and year). This suggests we don't have test262 coverage for the case where the year is out of range, but I'll try to write that test and will fix this in another PR after yours lands.

catamorphism reviewed (approved):
LGTM, thanks!

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -350,7 +350,7 @@ ISO8601::PlainDate TemporalCalendar::monthDayFromFields(JSGlobalObject* globalOb
     RETURN_IF_EXCEPTION(scope, { });
     double year = referenceYear.value_or(1972);
```
This should not be double

sosukesuzuki replied:
Fixed

sosukesuzuki committed:
[JSC] Replace `isValidISODate` with `isDateTimeWithinLimits` in `TemporalCalendar::monthDayFromFields`
https://bugs.webkit.org/show_bug.cgi?id=303163

Reviewed by Yusuke Suzuki.

The `isValidISODate` check after `regulateISODate` is redundant because:

- When overflow is `Reject`, `regulateISODate` already calls `isValidISODate`
  and returns `std::nullopt` if invalid.
- When overflow is `Constrain`, the month and day are clamped to valid ranges,
  so `isValidISODate` will always return true.

However, according to the spec, `monthDayFromFields` should call
`ISODateWithinLimits` to check that the year is within range.
This is necessary for future non-ISO calendar support where the
reference year may be invalid in rare cases.

See: https://github.com/tc39/proposal-temporal/issues/2997

* Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
(JSC::TemporalCalendar::monthDayFromFields):

Canonical link: https://commits.webkit.org/303878@main
